src/util/Generators.h

```diff
@@ -22,30 +24,67 @@ namespace ad_utility {
 // returns false. If the `aggregator` returns false, the cached value is
 // discarded. If the cached value is still present once the generator is fully
 // consumed, `onFullyCached` is called with the cached value.
-// NOTE: The `int` is just a dummy value.
+template <typename InputRange, typename AggregatorT, typename FullyCachedFuncT>
+class CachingWrapper
+    : public InputRangeFromGet<ql::ranges::range_value_t<InputRange>> {
+ public:
+  using T = ql::ranges::range_value_t<InputRange>;
+
+  CachingWrapper(InputRange generator, AggregatorT aggregator,
+                 FullyCachedFuncT onFullyCached)
+      : generator_(std::move(generator)),
+        aggregator_(std::move(aggregator)),
+        onFullyCached_(std::move(onFullyCached)) {}
+
+  std::optional<T> get() override {
+    if (!initialized_) {
+      it_ = generator_.begin();
+      initialized_ = true;
+    }
+
+    if (it_ == generator_.end()) {
+      if (shouldBeAggregated_ && aggregatedData_.has_value()) {
+        onFullyCached_(std::move(aggregatedData_.value()));
+      }
```

Comment:
```suggestion
      if (aggregatedData_.has_value()) {
        AD_CORRECTNESS_CHECK(shouldBeAggregated);
        onFullyCached_(std::move(aggregatedData_.value()));
      }
```
(a little more precise on how those are intertwined).